weld/llvm.rs

```diff
@@ -2384,6 +2384,96 @@ impl LlvmGenerator {
+                                                 print_type(&**elem_ty), print_type(&*param_ty));
+                            }
+                        }
+
+                        // Check that type of key is scalar
+                        let key_ll_sym;
+                        let key_ll_ty;
+                        let last_block = keyfunc.blocks.last().unwrap();
+                        if let Terminator::ProgramReturn(ref key_sym) = last_block.terminator {
+                            let key_ty = keyfunc.symbol_type(key_sym)?;
+                            if let Scalar(_) = *key_ty {
+                                self.gen_cmp(key_ty)?;
+                                let (key_ll_tyc, key_ll_symc) = self.llvm_type_and_name(keyfunc, key_sym)?;
+                                key_ll_ty = key_ll_tyc;
+                                key_ll_sym = key_ll_symc;
+                            } else {
+                                return weld_err!("Sort key function must have scalar return type");
+                            }
+                        } else {
+                            return weld_err!("Sort key Function must have return type");
+                        }
+
+                        // Generate prelude code
+                        let mut keyfunc_ctx = &mut FunctionContext::new(false);
+                        for (arg, ty) in keyfunc.locals.iter() {
+                            let arg_str = llvm_symbol(&arg);
+                            let ty_str = self.llvm_type(&ty)?;
+                            keyfunc_ctx.add_alloca(&arg_str, &ty_str)?;
+                        }
+                        for b in keyfunc.blocks.iter() {
+                            for s in b.statements.iter() {
+                                self.gen_statement(s, keyfunc, keyfunc_ctx)?
+                            }
+                        }
+
+                        // Add key function and sort prelude code
+                        let mut func_gen = IdGenerator::new("%keyfunc");
+                        let function_id = func_gen.next();
+                        let func_str = llvm_prefix(&&function_id);
+                        self.prelude_code.add(format!(
+                            "define {} {}({}) {{",
+                            key_ll_ty,
+                            func_str,
+                            str_args));
+                        self.prelude_code.add(keyfunc_ctx.alloca_code.result());
+                        self.prelude_code.add(keyfunc_ctx.code.result());
+                        self.prelude_code.add(format!("{}.ret = load {}, {}* {}", key_ll_sym, key_ll_ty, key_ll_ty, key_ll_sym));
+                        self.prelude_code.add(format!("ret {} {}.ret\n}}", key_ll_ty, key_ll_sym));
+                        let name = self.vec_names.get(elem_ty).unwrap();
+                        self.prelude_code.add(format!(
```

Comment:
if you do two different sorts on the same vector type you'll add this prelude code twice, which will cause an name conflict error

Reply:
Good call. I appended the key-function name {FUNC} as part of the sort function name in vector_sort.ll. So there can exist sort functions with the same type signatures, but they will differ since the key-function will be different. This also causes duplicated code being generated in cases where we call sort with the same key functions, but that should be a minor issue, right? I'm guessing that happens anyway due to function inlining.